### PR TITLE
Qodana simulation model

### PR DESCRIPTION
**Pull request by dariadiatlova**
**Description**: added small dataset preparation block, train, and evaluation module to train and run Roberta-model for solving a multilabel classification task

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,7 +1,7 @@
-setuptools==56.0.0
+setuptools~=47.3.2
```
Comment (nbirillo) [inline requirements.txt:1]: Why you decrease this version?

File: requirements.txt [hunk 2]
```diff
@@ -1,7 +1,7 @@
...
 
 # python code analysis tools
-pylint==2.7.4
+pylint~=2.5.3
```
Comment (nbirillo) [inline requirements.txt:4]: Why you decrease this version?

File: requirements.txt [hunk 3]
```diff
@@ -25,6 +25,9 @@ cohesion==1.0.0
 radon==4.5.0
 
 # extra libraries and frameworks
-django==3.2
-requests==2.25.1
+django~=3.0.8
+requests~=2.24.0
```
Comment (nbirillo) [inline requirements.txt:29]: Why you decrease these versions?

File: src/python/model/README.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
+# Qodana imitation model
```
Comment (nbirillo) [inline src/python/model/README.md:1]: Please, transfer your model into `qodana` directory and call it something like this: `initation_model`

File: src/python/model/README.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
+# Qodana imitation model 
+
+## Architecture 
+`RobertaForSequenceClassification` model with `BCEWithLogitsLoss` solve multilabel classification task.
```
Comment (nbirillo) [inline src/python/model/README.md:4]: Could you please add some link to a description of RobertaForSequenceClassification and BCEWithLogitsLoss?

File: src/python/model/README.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
+# Qodana imitation model 
+
+## Architecture
```
Comment (nbirillo) [inline src/python/model/README.md:3]: Please, add a description block before the Architecture block to describe what this model does in general

File: src/python/model/README.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
+# Qodana imitation model 
+
+## Architecture 
+`RobertaForSequenceClassification` model with `BCEWithLogitsLoss` solve multilabel classification task. 
+
+Model outputs is a tensor of size: `batch_size` x `num_classes`. Class probabilities are recieved by taking `sigmoid` and final predictions are computed by comparing the probability of each class with the `threshold`.
```
Comment (nbirillo) [inline src/python/model/README.md:6]: Please, add some description about batch_size and num_classes

Comment (nbirillo) [reply src/python/model/README.md:6]: Also you should describe what is `class`

File: src/python/model/README.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+`RobertaForSequenceClassification` model with `BCEWithLogitsLoss` solve multilabel classification task. 
+
+Model outputs is a tensor of size: `batch_size` x `num_classes`. Class probabilities are recieved by taking `sigmoid` and final predictions are computed by comparing the probability of each class with the `threshold`. 
+
+As classes might be unbalanced the used metric is `f1-score`.
+## What it does
+
+Model has two use cases:
+- It can be trained to predict a unique number of errors in a **block** of code, unfixed length. 
+- It can be trained to predict a unique number of errors in a **line** of code.
```
Comment (nbirillo) [inline src/python/model/README.md:13]: Could you please add some examples: code with founded inspection for each case?

File: src/python/model/README.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+As classes might be unbalanced the used metric is `f1-score`.
+## What it does
+
+Model has two use cases:
+- It can be trained to predict a unique number of errors in a **block** of code, unfixed length. 
+- It can be trained to predict a unique number of errors in a **line** of code. 
+___
+## Data preprocessing
+
+Please address to the [`following documentation`](https://github.com/hyperskill/hyperstyle/tree/roberta-model/src/python/evaluation/qodana) for labeling dataset and to the [`following documentation`](https://github.com/hyperskill/hyperstyle/tree/roberta-model/src/python/model/preprocessing) to preprocess data for model training and evaluation afterwards.
```
Comment (nbirillo) [inline src/python/model/README.md:17]: Please, use relevant paths everywhere instead of absolute ones

Comment (nbirillo) [reply src/python/model/README.md:17]: Could you please describe here how looks the final dataset for the model?

File: src/python/model/README.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+- It can be trained to predict a unique number of errors in a **block** of code, unfixed length. 
+- It can be trained to predict a unique number of errors in a **line** of code. 
+___
+## Data preprocessing
+
+Please address to the [`following documentation`](https://github.com/hyperskill/hyperstyle/tree/roberta-model/src/python/evaluation/qodana) for labeling dataset and to the [`following documentation`](https://github.com/hyperskill/hyperstyle/tree/roberta-model/src/python/model/preprocessing) to preprocess data for model training and evaluation afterwards. 
+
+___
+
+## How to train model
```
Comment (nbirillo) [inline src/python/model/README.md:21]: model -> the model

File: src/python/model/README.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+Run [`train.py`](https://github.com/hyperskill/hyperstyle/blob/roberta-model/src/python/model/train.py) script from the command line with the following arguments:
+
+Required arguments:
+
+`train_dataset_path`, `val_dataset_path` &#8209; path to the `train.csv` and `val.csv` dataset received by running [`split_dataset.py`](https://github.com/hyperskill/hyperstyle/blob/roberta-model/src/python/model/preprocessing/split_dataset.py) script.
```
Comment (nbirillo) [inline src/python/model/README.md:27]: Could you please add some description what are `train.csv` and `val.csv` here or into split_dataset.py directly

File: src/python/model/common/metric.py [hunk 12]
```diff
@@ -0,0 +1,21 @@
+import torch
+from sklearn.metrics import multilabel_confusion_matrix
+
+
+class Metric:
```
Comment (nbirillo) [inline src/python/model/common/metric.py:5]: Please, add some descriptions for this class. Maybe is it better to call it Measurer?
Moreover, it looks that you should have several fields in this class:
- threshold
- predictions
- targets
In init method you could get threshold and evaluation_predictions

- `get_f1_score` should be private (marked __)
- for `f1_score` in compute_metric result please create an enum class with keys of the mesuares

File: src/python/model/common/evaluation_config.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+                        type=int,
+                        default=40,
+                        help='Sequence length of 1 sample after tokenization, default is 40.')
+
+    parser.add_argument('-bs', '--batch_size',
+                        type=int,
+                        default=8,
+                        help='Batch_size – default is 8.')
+
+    parser.add_argument('-th', '--threshold',
```
Comment (nbirillo) [inline src/python/model/common/evaluation_config.py:34]: You have several same arguments in the train and evaluation config. Please, move them into constants with short names, long names, and descriptions.

File: src/python/model/common/train_config.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
+import argparse
+
+import torch
+from src.python.model.common.util import MarkingArgument
+from transformers import Trainer, TrainingArguments
+
+
+class MultilabelTrainer(Trainer):
```
Comment (nbirillo) [inline src/python/model/common/train_config.py:8]: Please, add some description for this class

File: src/python/model/common/util.py [hunk 15]
```diff
@@ -0,0 +1,20 @@
+from enum import Enum, unique
+
+
+@unique
+class MarkingArgument(Enum):
```
Comment (nbirillo) [inline src/python/model/common/util.py:5]: What means Marking? Can we rename it?

File: src/python/model/dataset/dataset.py [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    """ MarkingArgument.ID.value is a an id of the solution that corresponds to the line
+        MarkingArgument.INSPECTIONS.value is a is a target column name in dataset
+        ColumnName.CODE.value is an observation column name in dataset where lines of code are stored
+    """
+
+    def __init__(self, data_path: str, context_length: int):
+        super().__init__()
+        df = pd.read_csv(data_path)
+        tokenizer = RobertaTokenizer.from_pretrained('roberta-base')
+        code = list(map(str, df['code']))
```
Comment (nbirillo) [inline src/python/model/dataset/dataset.py:22]: We have constants for column names, please, use them

File: src/python/model/evaluation.py [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+from src.python.model.common.evaluation_config import configure_arguments
+from src.python.model.common.metric import Metric
+from src.python.model.common.util import MarkingArgument
+from src.python.model.dataset.dataset import QodanaDataset
+from src.python.review.common.file_system import Extension
+from torch.utils.data import DataLoader
+from transformers import RobertaForSequenceClassification
+
+
+def main():
```
Comment (nbirillo) [inline src/python/model/evaluation.py:18]: I think you should separate it into several functions

File: src/python/model/preprocessing/README.md [hunk 18]
```diff
@@ -0,0 +1,52 @@
+# Data preprocessing
+### Step 1
```
Comment (nbirillo) [inline src/python/model/preprocessing/README.md:2]: Please, add small description what this module does

File: src/python/model/preprocessing/encode_data.py [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+
+    parser.add_argument('-ohe', '--one_hot_encoding',
+                        help='If True, target column will be represented as one-hot-encoded vector. '
+                             'The length of each vector is equal to the unique number of classes. '
+                             'Default is True.',
+                        action='store_false')
+
+
+def __one_hot_encoding(df: pd.DataFrame) -> pd.DataFrame:
+    """ transform: ['1, 2', '3'] array([[1, 1, 0], [0, 0, 1]])
```
Comment (nbirillo) [inline src/python/model/preprocessing/encode_data.py:54]: Please, write more detailed description

File: src/python/model/preprocessing/encode_data.py [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+        lines_with_context = []
+        for current_line_index, current_line in enumerate(self.lines):
+            context = self.add_context_before(current_line_index, current_line)
+            context = self.add_context_after(context, current_line_index)
+            lines_with_context.append(context[0])
+        lines_with_context = pd.Series(lines_with_context)
+        self.df[ColumnName.CODE.value] = lines_with_context
+        return self.df
+
+    def add_context_before(self, current_line_index: int, current_line: str) -> List:
```
Comment (nbirillo) [inline src/python/model/preprocessing/encode_data.py:91]: Please, add some description

Comment (nbirillo) [reply src/python/model/preprocessing/encode_data.py:91]: Also, concretize the returned type. List of what? Use `List[another type]`

**Review by nbirillo — CHANGES_REQUESTED** [review]
Great, thanks!

File: src/python/model/common/metric.py (hunk 12 above)
Comment (dariadiatlova) [reply src/python/model/common/metric.py]: can I leave `get_f1_score` public as I use the method during [evaluation](https://github.com/hyperskill/hyperstyle/blob/roberta-model/src/python/evaluation/qodana/imitation_model/evaluation.py#L46) (model outputs differ on training and evaluation stage – I thought it would be good to separate computation into 2 parts)

Comment (dariadiatlova) [reply src/python/model/common/metric.py]: and I think I didn't get what are the **keys of the mesuares** :( 
could you explain, please?

firstly I thought it should be something like:
@unique
class MeasurerArgument(Enum):
    FP = 'false_positives'
    FN = 'false_negatives'
    TP = 'true_positives'
    TN = 'true_negatives'
but it looks like we compute values in method and there's no need to store just names in enum class

Comment (nbirillo) [reply src/python/model/common/metric.py]: Yes, I meant something like this for the returned dictionary in the compute_metric function:
`return {'f1_score': self.get_f1_score(predictions, torch.tensor(targets))}` -> `return {MeasurerArgument.F1_score.value: self.get_f1_score(predictions, torch.tensor(targets))}`

Then you can get necessary values from this dict by consts, not by strings

File: requirements.txt [hunk 21]
```diff
@@ -28,3 +28,6 @@ radon==4.5.0
 django==3.2
 requests==2.25.1
 argparse==1.4.0
+
+pytest~=6.2.3
```
Comment (nbirillo) [inline requirements.txt:32]: Why did you add it into the main requirements file?

File: requirements.txt [hunk 22]
```diff
@@ -28,3 +28,6 @@ radon==4.5.0
...
+jsonschema~=3.2.0
```
Comment (nbirillo) [inline requirements.txt:33]: And this

File: src/python/evaluation/qodana/imitation_model/common/evaluation_config.py [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+                        help='Path to the directory where labeled dataset will be saved. Default is the parent folder'
+                             'of test_dataset_path.')
+
+    parser.add_argument('-sf', '--save_f1_score',
+                        default=None,
+                        action="store_true",
+                        help=f'If enabled report with f1 scores by class will be saved to the {Extension.CSV.value}'
+                             ' File will be saved to the labeled dataset parent directory. Default is False.')
+
+    parser.add_argument(ModelCommonArguments.CONTEXT_LENGTH.value.short_name,
```
Comment (nbirillo) [inline src/python/evaluation/qodana/imitation_model/common/evaluation_config.py:31]: ModelCommonArguments -> ModelCommonArgument

File: src/python/evaluation/qodana/imitation_model/common/metric.py [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+from src.python.evaluation.qodana.imitation_model.common.util import MeasurerArgument
+
+logger = logging.getLogger(__name__)
+
+
+class Measurer:
+    def __init__(self, threshold: float):
+        self.threshold = threshold
+
+    def get_f1_score(self, predictions: torch.tensor, targets: torch.tensor) -> Union[float, None]:
```
Comment (nbirillo) [inline src/python/evaluation/qodana/imitation_model/common/metric.py:15]: Please, don't use Union[float, None]. Use Optional[float]

File: src/python/evaluation/qodana/imitation_model/common/util.py [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+    ID = 'id'
+    IN_ID = 'inspection_id'
+    INSPECTIONS = 'inspections'
+    INPUT_IDS = 'input_ids'
+    LABELS = 'labels'
+    DATASET_PATH = 'dataset_path'
+    STEPS = 'steps'
+    WEIGHTS = 'weights'
+    WANDB = 'wandb'
+    SEED = 42
```
Comment (nbirillo) [inline src/python/evaluation/qodana/imitation_model/common/util.py:17]: It is a bad practise use different types for values in Enum classes

**Review by nbirillo — CHANGES_REQUESTED** [review]
Great! I have several small comments

**Review by nbirillo — APPROVED** [review]
Great! Thank you!
